library/bignum_mod.c

```diff
@@ -209,7 +195,68 @@ int mbedtls_mpi_mod_modulus_setup( mbedtls_mpi_mod_modulus *m,
 /* END MERGE SLOT 6 */
 
 /* BEGIN MERGE SLOT 7 */
+int mbedtls_mpi_mod_read( mbedtls_mpi_mod_residue *r,
+                          const mbedtls_mpi_mod_modulus *m,
+                          const unsigned char *buf,
+                          size_t buflen,
+                          mbedtls_mpi_mod_ext_rep ext_rep )
+{
+    int ret = MBEDTLS_ERR_MPI_BAD_INPUT_DATA;
+
+
+    /* Do our best to check if r and m have been set up */
+    if( r->limbs == 0 || m->limbs == 0 )
+        goto cleanup;
+    if( r->limbs != m->limbs )
+        goto cleanup;
+
+    ret = mbedtls_mpi_mod_raw_read( r->p, m, buf, buflen, ext_rep );
+    if( ret != 0 )
+        goto cleanup;
+
+    r->limbs = m->limbs;
+
+    if( m->int_rep == MBEDTLS_MPI_MOD_REP_MONTGOMERY )
+       ret = mbedtls_mpi_mod_raw_to_mont_rep( r->p, m );
 
+cleanup:
+    return ( ret );
+}
+
+int mbedtls_mpi_mod_write( const mbedtls_mpi_mod_residue *r,
+                           const mbedtls_mpi_mod_modulus *m,
+                           unsigned char *buf,
+                           size_t buflen,
+                           mbedtls_mpi_mod_ext_rep ext_rep )
+{
+    int ret = MBEDTLS_ERR_MPI_BAD_INPUT_DATA;
+    int conv_ret = 0;
```

Comment:
I find the combinating of `ret` and `conv_ret` very confusing to read - what would it take to just use `ret`?

Reply:
I agree, we could use just `ret`. If it fails on the input checking it would be an `MBEDTLS_ERR_MPI_BAD_INPUT_DATA` but if it fails during convertion it will still return the appropriate ret from `mbedtls_mpi_mod_raw_from_mont_rep()` or `mbedtls_mpi_mod_raw_to_mont_rep()`

It should warrant a refactor by itself though, I can fix it on the following PR bringing in the automated tests.

Reply:
I thought about removing `conv_ret`, but I can't see a way to do without it. We need to consider both the return value of the write and the second conversion to decide what to return.

We could limit the scope of `conv_ret` though. Would you consider something like this an improvement?
https://github.com/yanesca/mbedtls/commit/84bee4c49230ecd50ab6105faf696ff015f03624

Reply:
(If `mbedtls_mpi_mod_raw_write()` fails and `mbedtls_mpi_mod_raw_to_mont_rep()` succeeds, we need to return the error from write. If both of them fail, we need to return the error from the conversion. I can't see a way to do this without saving both error codes.)